contracts/InternalMarket/InternalMarketBase.sol

```diff
@@ -156,30 +174,52 @@ contract InternalMarketBase {
         require(amount == 0, "InternalMarket: amount exceeds offer");
     }
 
-    function _withdraw(
+    function _mint(address to, uint256 amount) internal virtual {
+        // FIXME: ask marko if we can mint unwrapped tokens to investors.
+        tokenExternal.mint(address(this), amount);
+        tokenInternal.mint(to, amount);
+    }
+
+    function _deposit(address from, uint amount) internal virtual {
+        tokenExternal.transferFrom(from, address(this), amount);
+        tokenInternal.mint(from, amount);
+    }
+
+    function _matchOffer(
         address from,
         address to,
         uint256 amount
     ) internal virtual {
-        _beforeWithdraw(from, amount);
+        _beforeMatchOffer(from, to, amount);
         require(
-            daoToken.transfer(to, amount),
+            tokenInternal.transfer(to, amount),
+            "InternalMarketBase: transfer failed"
+        );
+        require(
+            exchangeToken.transferFrom(to, from, _convertToUSDC(amount)),
             "InternalMarketBase: transfer failed"
         );
     }
 
-    function _matchOffer(
+    function _withdraw(
         address from,
         address to,
         uint256 amount
     ) internal virtual {
-        _beforeMatchOffer(from, to, amount);
-        require(
-            daoToken.transfer(to, amount),
-            "InternalMarketBase: transfer failed"
-        );
+        if (
+            _shareholderRegistry.isAtLeast(
+                _shareholderRegistry.CONTRIBUTOR_STATUS(),
+                from
+            )
+        ) {
+            _beforeWithdraw(from, amount);
```

Comment:
What happens when a contributor is forced into investor role while having their tokens in the internal market?